keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java

```diff
@@ -350,4 +351,84 @@ private DoubleTensor pluckFromPartial(DoubleTensor from, int... indices) {
             .reshape(pluckShape);
     }
 
+    public DualNumber ifThenElse(BooleanTensor predicate, DualNumber els) {
+        if (predicate.allTrue()) {
+            return new DualNumber(this.value, this.getPartialDerivatives());
+        } else if (predicate.allFalse()) {
+            return new DualNumber(els.value, els.getPartialDerivatives());
+        } else {
+            return pluckFromMixedPredicate(predicate, this, els);
+        }
+    }
+
+    private DualNumber pluckFromMixedPredicate(BooleanTensor predicate, DualNumber thn, DualNumber els) {
+        int[] thenShape = thn.value.getShape();
+        double[] flatPredicate = predicate.asFlatDoubleArray();
+        Map<Long, List<DoubleTensor>> ifAndElseDualNumbers = new HashMap<>();
+
+        for (int i = 0; i < flatPredicate.length; i++) {
+            boolean condition = flatPredicate[i] == 1.0;
```

Comment:
`condition` can be inlined - only used once and doesn't improve the explanatory power of the code